sound/soc/sof/topology.c

```diff
@@ -3018,8 +3017,27 @@ int snd_sof_complete_pipeline(struct snd_sof_dev *sdev,
 	ret = sof_ipc_tx_message(sdev->ipc,
 				 ready.hdr.cmd, &ready, sizeof(ready), &reply,
 				 sizeof(reply));
-	if (ret < 0)
-		return ret;
+
```

Comment:
it's not clear to me why there is even a need to send a PIPE_COMPLETE IPC command if the topology failed earlier. That looks like fuzzing the DSP by sending commands that are known to fail already. Fuzzing is good but not as part of the error handling.

Reply:
Anyway it's not a small change and it'd need to come with broken topologies and test cases where we know what needs to work and what doesn't.

Reply:
PIPELINE_COMPLETE is sent at the end of topology parsing as the topology file format does not guarantee any order of component or DAI definitions that are parsed by the driver and FW. They can be completely random. PIPELINE_COMPLETE is only sent after all components/DAIs have been parsed to tell FW "there's no more components coming for pipe X, please try and realize pipe X now"

Agree, it will need testing with fuzzer and broken topologies. We can close this until fuzzer is ready with incomplete pipelines. @ranj063 fyi.